ec/ecresource/deploymentresource/topology_dedicated_masters.go

```diff
@@ -0,0 +1,365 @@
+		case "warm":
+			rawTopology = esPlan.WarmTier
+		case "cold":
+			rawTopology = esPlan.ColdTier
+		case "frozen":
+			rawTopology = esPlan.FrozenTier
+		}
+
+		if rawTopology.IsNull() || rawTopology.IsUnknown() {
+			continue
+		}
+
+		size, zoneCount := getSizeAndZoneCount(ctx, topologyId, rawTopology, template)
+
+		// Calculate if there are >1 nodes in each zone:
+		// If the size is > the maximum size allowed in the zone, more nodes are added to reach the desired size.
+		instanceConfig := getTemplateInstanceConfiguration(template, topologyId)
+		maxSize := getMaxSize(instanceConfig)
+		var nodesPerZone int32
+		if size < maxSize || maxSize == 0 {
+			nodesPerZone = 1
+		} else {
+			nodesPerZone = size / maxSize
+		}
+
+		if size > 0 && zoneCount > 0 {
+			nodesInDeployment += zoneCount * nodesPerZone
+		}
+	}
+
+	return nodesInDeployment
+}
+
+func getSizeAndZoneCount(
+	ctx context.Context,
+	topologyId string,
+	rawTopology types.Object,
+	template models.DeploymentTemplateInfoV2,
+) (int32, int32) {
+	var topology es.ElasticsearchTopologyTF
+	diags := rawTopology.As(ctx, &topology, objectAsOptions)
+	if diags.HasError() {
+		tflog.Debug(ctx, "getSizeAndZoneCount: Failed to read topology.", withDiags(diags))
+		return 0, 0
+	}
+
+	var size int32
+	if !topology.Size.IsUnknown() && !topology.Size.IsNull() {
+		var err error
+		size, err = deploymentsize.ParseGb(topology.Size.ValueString())
+		if err != nil {
+			tflog.Debug(ctx, "getSizeAndZoneCount: Failed to parse topology size.", withError(err))
+			return 0, 0
+		}
+	} else {
+		// Fall back to template value
+		size = getTopologySize(template, topologyId)
+	}
+
+	var zoneCount int32
+	if !topology.ZoneCount.IsUnknown() && !topology.ZoneCount.IsNull() {
+		zoneCount = int32(topology.ZoneCount.ValueInt64())
+	} else {
+		// Fall back to template value
+		zoneCount = getTopologyZoneCount(template, topologyId)
+	}
+
+	return size, zoneCount
+}
+
+// If the instance-config-id is set in the topology, loads that specific IC via API
+// If no instance-config-id is set, uses the IC set in the deployment template
+func getInstanceConfiguration(
+	ctx context.Context,
+	rawTopology types.Object,
+	deploymentInstanceConfigs []models.InstanceConfigurationInfo,
+) *models.InstanceConfigurationInfo {
+
+	if rawTopology.IsUnknown() || rawTopology.IsNull() {
+		return nil
+	}
+
+	var topology es.ElasticsearchTopologyTF
+	diags := rawTopology.As(ctx, &topology, objectAsOptions)
+	if diags.HasError() {
+		tflog.Debug(ctx, "getInstanceConfigurationId: Failed to read topology.", withDiags(diags))
+		return nil
+	}
+
+	if topology.InstanceConfigurationId.IsUnknown() || topology.InstanceConfigurationId.IsNull() {
+		return nil
+	}
+
+	icId := topology.InstanceConfigurationId.ValueStringPointer()
+	if icId == nil || *icId == "" {
+		return nil
+	}
+
+	var deploymentIc *models.InstanceConfigurationInfo
+	for _, ic := range deploymentInstanceConfigs {
+		if ic.ID == *icId {
```

Comment:
Shall we take `ConfigVersion` into consideration?

Reply:
Yeah it is safer. Currently the list of ICs will always contain just one version, but if that part of the API ever changes we still get the right one.

Reply:
Added with the latest commit